src/libs/kata-sys-util/src/hooks.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+// Copyright (c) 2019-2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::hash::{Hash, Hasher};
+use std::io::{self, Read, Result};
+use std::path::Path;
+use std::time::Duration;
+
+use subprocess::{ExitStatus, Popen, PopenConfig, PopenError, Redirection};
+
+use crate::{eother, sl};
+
+const DEFAULT_HOOK_TIMEOUT_SEC: i32 = 10;
+
+#[derive(Debug, Default, Clone)]
+struct HookKey(oci::Hook);
```

Comment:
ooi, why can't you just `#[derive(Hash)` on `oci::Hook` itself to avoid having to introduce `HookKey`?

Reply:
oci::Hook is an auto-generated structure, it may bring some maintenance burden to add "Eq, Hash" to the oci::Hook.
And it's possible to switch from agent/oci to the published oci-spec crate, to reduce maintenance burden:)

Reply:
It's hybrid, auto-generated and then modified manually.